subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java

```diff
@@ -40,13 +47,48 @@ public AbstractDaemonCompiler(File daemonWorkingDir, Compiler<T> delegate, Compi
     @Override
     public WorkResult execute(T spec) {
         DaemonForkOptions daemonForkOptions = toDaemonOptions(spec);
-        CompilerDaemon daemon = compilerDaemonFactory.getDaemon(daemonWorkingDir, daemonForkOptions);
-        CompileResult result = daemon.execute(delegate, spec);
+        WorkerDaemon daemon = compilerDaemonFactory.getDaemon(CompilerDaemonServer.class, daemonWorkingDir, daemonForkOptions);
+        WorkerDaemonResult result = daemon.execute(adapter(delegate), spec);
         if (result.isSuccess()) {
             return result;
         }
         throw UncheckedException.throwAsUncheckedException(result.getException());
     }
 
+    private CompilerWorkerAdapter<T> adapter(Compiler<T> compiler) {
+        return new CompilerWorkerAdapter<T>(compiler);
+    }
+
     protected abstract DaemonForkOptions toDaemonOptions(T spec);
+
+    private static class CompilerWorkerAdapter<T extends CompileSpec> implements WorkerDaemonAction<T> {
+        private final Compiler<T> compiler;
+
+        CompilerWorkerAdapter(Compiler<T> compiler) {
+            this.compiler = compiler;
+        }
+
+        @Override
+        public WorkerDaemonResult execute(T spec) {
+            return new WorkerDaemonResult(compiler.execute(spec).getDidWork(), null);
+        }
+
+        @Override
+        public String getDescription() {
+            return compiler.getClass().getName();
+        }
+    }
+
+    // TODO Come up with a better way to set up the worker implementation classpath
+    // This is a hack to get the appropriate classpath on the worker implementation classpath for compiler daemons.
+    // The classpath is derived from the implementation class and when this is WorkerDaemonServer, we get the classpath
+    // from the classloader in the Gradle Core API classloader scope (which only contains certain jars).  Using this
+    // class causes the classpath to be inferred from the Gradle API scope classloader instead so that we get the necessary
+    // jars for a compiler daemon.
+    public static class CompilerDaemonServer extends WorkerDaemonServer {
```

Comment:
Compiler daemons uses the **internal** worker daemons API.
If they used the **public** worker daemons API we could gather the correct classpath here and use `WorkerDaemonExecutor.classpath(..)` directly. This should achieve the same thing, right?